### ISSUE TITLE
Float secondary index - testing bulk insertion and range queries

### ISSUE THREAD

**Issue by healiseu** [RedisLabsModules/RediSearch#493]
UPDATE: [New code][5] to cover many subtleties on the issue - Read a summary [here][6]
----------------------------------------------------------------------------------------------------------------------------------------
Hi, first thanks for sharing openly this project, I would like to start experimenting with your search engine to see how I can embed it in [TRIADB][1]. I have already built an implementation that uses Josiah Carlson Redis Object-Mapper ([ROM][2]) but it's rather slow on inserts and it has limited functionality on queries.

Nevertheless TRIADB is a Python Database Framework and it is very convenient for developers to use an Object Mapper like ROM or Walrus or Stdnet, something that can be developed on top of redisearch-py I think. In particular I am interested in Unique indexes, composite indexes and ManyToMany relationships. This is an essential and basic functionality for TRIADB.

To be more specific for a start I want to use redisearch to measure the memory footprint and the insertion speed for the values of a column from a table. Each column of data in TRIADB belongs to a data type and each data type stores distinct, i.e. single instance, values of a specific format e.g. integers, floats, dates, categorical data, etc. Therefore I want to learn what is the most efficient way, speed is No1 priority and memory size is second, to create and apply a unique index constraint in RediSearch engine.

For example I wrote this simple test [demo][3] to enter 1 million sorted floats in the range [-5000, 5000]. 

1. How is it possible to specify the representation of a number in redisearch ? Does it account internally for differences between [numeric types][4]
2. What is the optimum value for the chunk_size, what is a general guideline ?
3. How the memory sizes reported from info(), compare to those reported from redis client ?  The `info memory` command in my system shows:
   a.  used_memory_human  :  256 MB
   b.  used_memory_dataset :  209 MB
4. What is the most efficient way to build a unique constraint index apart from using Python try-except block ?
5. What other ways would you think in order to optimize data insertion, a bitmap index for primary key auto-increment perhaps ?

[1]: http://healis.eu/triadb/
[2]: https://github.com/josiahcarlson/rom
[3]: https://gist.github.com/healiseu/549a32c35b2b45940147fc4c5dcb88a4
[4]: https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html
[5]: https://gist.github.com/healiseu/0856470814b87f78d188dd1ec82d712b
[6]: https://github.com/RedisLabsModules/RediSearch/issues/493#issuecomment-422552591

**Comment by mnunberg** [comment]
I am looking at your example, so `NumericFilter` is a (deprecated) argument for search that applies the constraints at query-time. In general, most of RediSearch's search/sort logic applies at query time. The only kind of sorting performed at indexing time is the inherent sort-by-id.

1. Internally every number is treated as a double, however, when storing the number in the index, the storage is optimized depending on the specific width of the number, and whether or not it is integral. Generally using floating non-integral numbers will use the most memory, while using integers will use less memory, the smaller the number, the smaller the footprint.
2. You'll need to play around with `chunk_size` and see what's faster. You can also attain higher throughput (potentially) if you split the workload into multiple TCP connections (i.e. multiple clients). In Python this is a bit tricky because Python doesn't actually do proper threading, so you might make use of the built-in `multiprocessing` module.
3. The `info` reports the size of the indexes, whereas `info memory` reports the size of everything stored in Redis (including the pointer data itself, etc.).
4. The primary key is the quickest. If you know the keys you are using before hand, you could perhaps use Redis' `EXISTS` command to verify. `EXISTS` can check multiple keys, but unfortunately won't tell you which ones exist and which ones don't.
5. In your case, if you know how to use a profiler, you could see what functions in Redis (or RediSearch) are occupying the most insertion time. I (or you even!) could analyze it and see what is causing slowdowns, if any.

**Comment by healiseu** [comment]
Hi Mark, thanks for your prompt reply to my questions, here is some more detail and feedback in what exactly I am asking.
1. There is a huge difference calculating memory size between numpy array structure and redis i.e. 8bytes x 1m floats = 7.6MB and redisearch used_memory = 256MB. There is also at least 100MB discrepancy between sizes reported with redis-cli and those from python client.info().
2. I played with chunk_size=100, (117sec, 78sec), chunk_size=10000, (131sec, 72sec) and chunk_size=100000, (KeyboardInterrupt due to excessive long time ). You have to note also that there is a significant difference between the elapsed time and the CPU process time. Why is that ?
3. I started with a clean instance of redis, i.e.  used_memory = 1MB and then run the demo code
4. OK I want something that I can call at  a higher level, e.g. from python like Float(unique=True, index=True)
5. Sure, but before I decide to go in depth I am trying to understand why is there such a huge difference both in terms of memory consumption and speed in comparison with a memory data structure like numpy array.

In conclusion, although RediSearch massive insertion is faster than ROM (see also [this test][1]), they are both very slow in comparison with in-memory data structures e.g. NumPy, Pandas, and/or other BI systems like qlikview. And generally speaking I would like to scale horizontally when I have exhausted my current hardware resources.

[1]: https://gist.github.com/healiseu/df9f04907b12ad365bb9d7dfaa347a47

**Comment by mnunberg** [comment]
1. The `client.info` uses `FT.INFO`, not the redis memory statistics, which is what you're running.
2. CPU time refers to the *client* CPU time. If the client is simply sitting there, waiting for the server to process things, then the client is not consuming a whole lot of CPU time, but the wall-time (i.e. 'elapsed time') is certainly advancing.
3. ok
4. We don't support unique values except at the key/index level. I suppose you can file a feature request for this.
5. A numpy array is simply an array of 8-byte (or 4 byte?) floats or doubles back to back in memory. It exists as a single chunk, and is addressable by its index only. If  you multiply 8bytes by 1M, you'll end up with about 8MB (a bit less, as you can see). RediSearch on the other hand is an inverted index. It is a document/value model, where documents have primary keys, and then a bunch of secondary key-values. RediSearch allows you to search and retrieve documents by their values. This is something you cannot do in numpy. If you have a need to search for "which position is this value located at", you'll find yourself performing many lookups (and the number of lookups is slower the more values are added).

    You should note that there is an overhead of about 32 bytes just for *metadata* of the document itself, so 32 * 100000M is 32M. Then, there is the index structure itself, which grows depending on how unique your values are, then there is the actual storing of that document in Redis, which comes with the tax of the ID string and the document entry in Redis itself, and so on.

We offer a (commercial) product that allows you to scale horizontally, either on the cloud, or deployed in your environment, which runs a cluster-capable version of redisearch.

**Comment by healiseu** [comment]
Summary

**Comment by healiseu** [comment]
The latency on server round trip time and the differences on memory size and insertion speed in all three cases I described are significant because the sample size, i.e. volume of data,  is small, only 250k floats. Redis vanilla, RediSearch query performance seems satisfactory although I would like to run the test on higher volumes of data with multiple secondary indexes. Nevertheless, as I have demonstrated, the insertion speed for RediSearch, let aside the memory size, is unacceptable for that sample size of data and prevents me from using/testing with TRIADB.

**Comment by tw-bert** [comment]
@healiseu : nice benchmarks. I've been meaning to participate in this thread as well, but didn't find the time this week.

For your case number 3:

You might be able to increase OPS by batching using Lua. 
You might be able to reduce memory footprint by using RediSearch `NOSAVE`.

**Comment by healiseu** [comment]
@tw-bert : thank you, I have tons of things I am experimenting with and that makes it extremely difficult to get into details with each one of them. For this specific experiment, case 3 with RedisSearch, if you are willing to write down some lua code I can try to execute it using [redis-py lua scripting][1] Moreover if you also guide me on how to tweak Redis server I will be happy to make public another benchmark based on my machine. But you said, you "might be able" to increase OPS, i.e. drop 17sec down to 5sec ???  and that leaves out memory optimization of course.

[1]: https://github.com/andymccurdy/redis-py#lua-scripting

**Comment by tw-bert** [comment]
@healiseu : Yes sure, no problem and interesting for us as well. Can you update your gist, so it reflects your new insights [here](https://github.com/RedisLabsModules/RediSearch/issues/493#issuecomment-423824059) ?

I can't predict any OPS increase, will have to test. But we've seen quite serious OPS improvements in the past, like (oldy) the first post in [this](https://groups.google.com/d/topic/redis-db/D4V6kDJNDsI/discussion) thread.

**Comment by healiseu** [comment]
@tw-bert : I am glad you are interested in my results. I have used Pandas in the past but recently I started digging into its foundation, NumPy, and its future i.e. Apache Arrow. These can be leveraged easily, i.e. functional, vector programming, to build very powerful staff. 

Generally speaking, from my perspective, the subject of how to build and use efficiently secondary indexes has to be elevated to **a first class citizen on database technology** and open-source libraries have to be built for this. Something similar to the higher-level data structures that Salvatore built for Redis is greatly missing I believe. 

Many open-source databases provide APIs on how to build indexes but these are limited in functionality and/or have complex interfaces. There are very few out there that realize this. The relatively recent resurgence of column-oriented databases, including time-series, and development of search engines, prove this is the case and the trend for the future. 

There are many things I investigate with TRIADB regarding how you construct these indexes, but most important **how you use them in row/table reconstruction, graph transformations and queries (like the one in this exercise)**. My plan is to explain these for your audience by [demonstrating TRIADB with Redis][1] on a pilot study with more than one tables linked on a hypergraph and more interesting-complex queries.

[1]: https://www.papercall.io/talks/90562/children/90563

**Comment by tw-bert** [comment]
@healiseu Interesting. But, can you follow up my comment (provide an updated gist so I can reproduce your benchmark)?

**Comment by healiseu** [comment]
Sorry, I cannot make public **now** any of my current work that is related to NumPy and PyArrow. I will disclose more information in due time, thank you again for your interest.

**Comment by tw-bert** [comment]
@healiseu You asked me some stuff;

> if you are willing to write down some lua code

and

>  i.e. drop 17sec down to 5sec ??? 

I offered to help, but with comments like the two above, I fear that I'm quickly loosing interest.

**Comment by healiseu** [comment]
@mnunberg for you interest I discovered that it seems there is a problem of redis-py pipeline handling 1,000,000 rows of duplets of an integer and a float value. I wrote a piece of code so that they can try to reproduce the error and I also pointed out that this is an issue with redis-py client.

Regarding to the performance  of bulk insertion it took almost 10sec in my environment to load the data using redis pipeline. Generally speaking for a TCP based client with a loopback interface in the same physical machine with the server, with the specs I gave you for my environment and the specific volume of data I run the test this should be in the range of (msec).

[1]: https://github.com/andymccurdy/redis-py/issues/850

**Comment by tw-bert** [comment]
@healiseu It might help to know that redis-py pipelines are no true pipelines. It uses the redis pipeline protocol, but everything is buffered client-side until execute. So no real fifo. See my discussion with Andy a few years back (to my knowledge this hasn't changed since), [here](https://groups.google.com/d/topic/redis-db/D4V6kDJNDsI/discussion) - (same link as my previous [comment](https://github.com/RedisLabsModules/RediSearch/issues/493#issuecomment-424232661)).
